### PR TITLE
Add several new methods to Types and Methods Assertions

### PR DESCRIPTION
**Pull request by fshchudlo**
Hello there.
I've added a couple of methods to types selectors and assertions which, I hope, you find useful.
Corresponding unit tests, of course, added. 

Approval tests are definitely broken since I've added new methods, but I'm not sure if I should change them since they are _approval_ tests. I would be grateful if you instruct me on this

## IMPORTANT 

* [x] The code complies with the [Coding Guidelines for C#](https://www.csharpcodingguidelines.com/).
* [x] The changes are covered by a new or existing set of unit tests which follow the Arrange-Act-Assert syntax such as is used [in this example](https://github.com/fluentassertions/fluentassertions/blob/daaf35b9b59b622c96d0c034e8972a020b2bee55/Tests/FluentAssertions.Shared.Specs/BasicEquivalencySpecs.cs#L33).
* [x] If the contribution affects [the documentation](https://github.com/fluentassertions/fluentassertions/tree/master/docs/_pages), please include your changes in this pull request so the documentation will appear on the [website](https://www.fluentassertions.com).

### CODE REVIEW

**Comment by fshchudlo** [comment]
@jnyrup @dennisdoomen, hi there
I've merged the conflict which I did not see when creating this PR. If I should do anything else to make this PR appropriate, reach out to me with any comments.

**Comment by dennisdoomen** [comment]
Sorry, completely missed this one. Will take a look tomorrow.

File: Src/FluentAssertions/TypeEnumerableExtensions.cs [hunk 1]
```diff
@@ -77,5 +77,60 @@ public static IEnumerable<Type> ThatImplement<T>(this IEnumerable<Type> types)
         {
             return new TypeSelector(types).ThatImplement<T>();
         }
+
+        /// <summary>
+        /// Filters to only include types that are classes.
+        /// </summary>
+        public static IEnumerable<Type> ThatAreClasses(this IEnumerable<Type> types)
```
Comment (dennisdoomen) [inline Src/FluentAssertions/TypeEnumerableExtensions.cs:84]: None of the these new members are covered by tests.

File: Src/FluentAssertions/TypeEnumerableExtensions.cs [hunk 2]
```diff
@@ -77,5 +77,60 @@ public static IEnumerable<Type> ThatImplement<T>(this IEnumerable<Type> types)
...
+        {
+            return new TypeSelector(types).ThatAreClasses();
+        }
+
+        /// <summary>
+        /// Filters to only include types that are not classes.
+        /// </summary>
+        public static IEnumerable<Type> ThatAreNotClasses(this IEnumerable<Type> types)
+        {
+            return new TypeSelector(types).ThatAreClasses();
```
Comment (dennisdoomen) [inline Src/FluentAssertions/TypeEnumerableExtensions.cs:94]: Shouldn't be using `ThatAreNotClasses`. Same for the other negated versions.

File: Src/FluentAssertions/Types/MethodInfoSelector.cs [hunk 3]
```diff
@@ -128,6 +128,17 @@ public MethodInfoSelector ThatAreNotDecoratedWithOrInherit<TAttribute>()
             return this;
         }
 
+        /// <summary>
+        /// Select return types of the methods
+        /// </summary>
+        /// <returns></returns>
+        public TypeSelector ReturnTypes()
```
Comment (dennisdoomen) [inline Src/FluentAssertions/Types/MethodInfoSelector.cs:135]: Can you give an example of why this would be useful?

File: Src/FluentAssertions/Types/TypeSelectorAssertions.cs [hunk 4]
```diff
@@ -292,6 +292,43 @@ public AndConstraint<TypeSelectorAssertions> NotBeDecoratedWithOrInherit<TAttrib
...
+        /// <param name="becauseArgs">
+        /// Zero or more objects to format using the placeholders in <paramref name="because" />.
+        /// </param>
+        public AndConstraint<TypeSelectorAssertions> BeSealed(string because = "", params object[] becauseArgs)
+        {
+            var notSealedTypes = Subject.Where(type => !type.IsCSharpSealed()).ToArray();
+            Execute.Assertion.ForCondition(!notSealedTypes.Any())
+                .BecauseOf(because, becauseArgs)
+                .FailWith("Expected all types to be sealed{reason}, but the following types are not:{0}{1}.", Environment.NewLine, GetDescriptionsFor(notSealedTypes));
+            return new AndConstraint<TypeSelectorAssertions>(this);
```
Comment (dennisdoomen) [inline Src/FluentAssertions/Types/TypeSelectorAssertions.cs:311]: (Nitpicking) We prefer an empty line before and after a multi-line statement like this.

File: Tests/FluentAssertions.Specs/Types/MethodInfoSelectorAssertionSpecs.cs [hunk 5]
```diff
@@ -231,5 +232,101 @@ public void When_asserting_methods_are_not_decorated_with_attribute_but_they_are
...
+            // Act
+            Action act = () =>
+                methodSelector.Should().Be(CSharpAccessModifier.Public);
+
+            // Assert
+            act.Should().NotThrow();
+        }
+
+        [Fact]
+        public void When_asserting_methods_does_not_have_specified_accessor_it_should_throw()
```
Comment (dennisdoomen) [inline Tests/FluentAssertions.Specs/Types/MethodInfoSelectorAssertionSpecs.cs:253]: I would rephrase tests like this `When_not_all_methods_have_the_specified_access_modifier_it_should_throw`. It avoids the repetitive `When_asserting_methods` and keeps them functional.

File: Tests/FluentAssertions.Specs/Types/MethodInfoSelectorAssertionSpecs.cs [hunk 6]
```diff
@@ -231,5 +232,101 @@ public void When_asserting_methods_are_not_decorated_with_attribute_but_they_are
...
+        {
+            // Arrange
+            var methodSelector = new MethodInfoSelector(typeof(ClassWithNonPublicMethods));
+
+            // Act
+            Action act = () =>
+                methodSelector.Should().Be(CSharpAccessModifier.Public);
+
+            // Assert
+            act.Should().Throw<XunitException>();
```
Comment (dennisdoomen) [inline Tests/FluentAssertions.Specs/Types/MethodInfoSelectorAssertionSpecs.cs:263]: We generally assert at least the relevant parts of the exception message to make sure the test fails for the right reasons.

**Comment by dennisdoomen** [comment]
> dennisdoomen approved these changes 12 minutes ago

Sorry, I meant to click the "Request Changes" button.

> Approval tests are definitely broken since I've added new methods, but I'm not sure if I should change them since they are approval tests. I would be grateful if you instruct me on this

You can run `AcceptApiChanges.ps1` to accept the changes and update the contract files.

Oh, and would you also be willing to update the docs and release notes under `docs\pages`?

**Comment by fshchudlo** [comment]
Sure. it's a pleasure for me to make a contribution to such a nice library :)
I'll do everything needed ASAP.

File: Src/FluentAssertions/TypeEnumerableExtensions.cs (hunk 2 above)
Comment (fshchudlo) [reply Src/FluentAssertions/TypeEnumerableExtensions.cs:94]: Fixed

File: Src/FluentAssertions/Types/TypeSelectorAssertions.cs (hunk 4 above)
Comment (fshchudlo) [reply Src/FluentAssertions/Types/TypeSelectorAssertions.cs:311]: Fixed

File: Src/FluentAssertions/TypeEnumerableExtensions.cs (hunk 1 above)
Comment (fshchudlo) [reply Src/FluentAssertions/TypeEnumerableExtensions.cs:84]: I'm a little bit confused since old members are also not covered with tests. 
Moreover, I see that all members of this class don't have any logic and they're just proxies to types that are covered by tests.
Are you sure they should be covered by tests? As I see such tests will inevitably be a copy-paste of other tests or such brittle things like Moqs `WasCalled` assertions

File: Src/FluentAssertions/Types/MethodInfoSelector.cs (hunk 3 above)
Comment (fshchudlo) [reply Src/FluentAssertions/Types/MethodInfoSelector.cs:135]: Sure. It's some kind of pseudocode

All of the methods that return collections should return immutable collections
`typeof(IDataReader).
        .Methods()
        .ReturnTypes()
        .ThatImplement<IEnumerable>()
        .Should()
        .BeImmutableCollection()`

All of the methods should return immutable types
`typeof(IDataReader).
        .Methods()
        .ReturnTypes()
        .Properties()
        .Should()
        .NotBeWritable()`

File: Src/FluentAssertions/TypeEnumerableExtensions.cs (hunk 1 above)
Comment (dennisdoomen) [reply Src/FluentAssertions/TypeEnumerableExtensions.cs:84]: You're right about the old code. We must have missed that in the past. But I spotted a bug in them during the review, which should have been caught by the tests.

Comment (fshchudlo) [reply Src/FluentAssertions/TypeEnumerableExtensions.cs:84]: Got it. I'll think about how to make tests that will not be very brittle.

File: Tests/FluentAssertions.Specs/Types/MethodInfoSelectorAssertionSpecs.cs (hunk 5 above)
Comment (fshchudlo) [reply Tests/FluentAssertions.Specs/Types/MethodInfoSelectorAssertionSpecs.cs]: Fixed

File: Tests/FluentAssertions.Specs/Types/MethodInfoSelectorAssertionSpecs.cs (hunk 6 above)
Comment (fshchudlo) [reply Tests/FluentAssertions.Specs/Types/MethodInfoSelectorAssertionSpecs.cs]: Fixed

**Comment by fshchudlo** [comment]
> Oh, and would you also be willing to update the docs and release notes under `docs\pages`?
I've described new `ReturnTypes`, `UnwrapTaskTypes` and `UnwrapEnumerableTypes`. Everything else is small changes that fit the description at the end of the docs: "There are so many possibilities and specialized methods that none of these examples do them good. Check out the TypeAssertionSpecs.cs".

Can you also instruct me about release notes? Should it be described as `5.11.0` or `6.1.0` version?

I need more time to add tests for `TypeEnumerableExtensions`. I hope I'll be able to add them tomorrow.

Also, I would be grateful if you instruct me about the need to rebase all these changes to a single commit since there are a lot of them now.

**Comment by dennisdoomen** [comment]
> I've described new `ReturnTypes`, `UnwrapTaskTypes` and `UnwrapEnumerableTypes`. Everything else is small changes that fit the description at the end of the docs

Looks good to me.

> Can you also instruct me about release notes? Should it be described as `5.11.0` or `6.1.0` version?

It's all going to be part of 6.0.0. Just add it under the "new" section.

> And the last one, I would be grateful if you instruct me about the need to rebase all these changes to a single commit since there are a lot of them now.

If you have not "designed" your PR as clean separated commits, will just squash-merge your changes. I myself follow a pretty rigorous practice to get a [clean source control history](https://www.continuousimprover.com/2020/03/keep-source-control-history-clean.html).

File: Src/FluentAssertions/TypeEnumerableExtensions.cs (hunk 1 above)
Comment (fshchudlo) [reply Src/FluentAssertions/TypeEnumerableExtensions.cs:84]: Fixed

**Comment by fshchudlo** [comment]
Well, as far as I can see, everything is done :)

But the build is broken for some reason not related to my changes, as I can see. Could you please look what’s the matter for it?

File: Src/FluentAssertions/TypeEnumerableExtensions.cs [hunk 7]
```diff
@@ -77,5 +77,60 @@ public static IEnumerable<Type> ThatImplement<T>(this IEnumerable<Type> types)
...
+        public static IEnumerable<Type> ThatAreNotClasses(this IEnumerable<Type> types)
+        {
+            return new TypeSelector(types).ThatAreNotClasses();
+        }
+        /// <summary>
+        /// Filters to only include types that are static classes.
+        /// </summary>
+        public static IEnumerable<Type> ThatAreStaticClasses(this IEnumerable<Type> types)
+        {
+            return new TypeSelector(types).ThatAreStaticClasses();
```
Comment (jnyrup) [inline Src/FluentAssertions/TypeEnumerableExtensions.cs:101]: Should these just be named `ThatAre[Not]Static()`?
That's what the implementation seems to do.

File: Src/FluentAssertions/TypeEnumerableExtensions.cs [hunk 8]
```diff
@@ -77,5 +77,60 @@ public static IEnumerable<Type> ThatImplement<T>(this IEnumerable<Type> types)
...
+        /// <summary>
+        /// Filters to only include types that satisfies the <paramref name="predicate"/> passed.
+        /// </summary>
+        public static IEnumerable<Type> ThatAre(this IEnumerable<Type> types, Func<Type, bool> predicate)
+        {
+            return new TypeSelector(types).ThatAre(predicate);
+        }
+
+        /// <summary>
+        /// Returns T for the types which are Task&lt;T&gt;; the type itself otherwise
```
Comment (jnyrup) [inline Src/FluentAssertions/TypeEnumerableExtensions.cs:121]:
```suggestion
        /// Returns T for the types which are <see cref="Task{T}" />; the type itself otherwise
```

File: Src/FluentAssertions/Types/TypeSelector.cs [hunk 9]
```diff
@@ -165,6 +167,96 @@ public TypeSelector ThatAreNotUnderNamespace(string @namespace)
...
+        }
+
+        /// <summary>
+        /// Returns T for the types which are Task&lt;T&gt;; the type itself otherwise
+        /// </summary>
+        public TypeSelector UnwrapTaskTypes()
+        {
+            types = types.Select(type =>
+            {
+                if (type.IsGenericType && type.GetGenericTypeDefinition() == typeof(Task<>))
```
Comment (jnyrup) [inline Src/FluentAssertions/Types/TypeSelector.cs:221]: Can this also handle `ValueTask`?

File: Src/FluentAssertions/Types/TypeSelector.cs [hunk 10]
```diff
@@ -165,6 +167,96 @@ public TypeSelector ThatAreNotUnderNamespace(string @namespace)
...
+                    if (type.IsGenericType)
+                    {
+                        if (type.GetGenericTypeDefinition() == typeof(IEnumerable<>))
+                        {
+                            return type.GetGenericArguments().Single();
+                        }
+
+                        var iEnumerableDefinition = type.GetInterfaces()
+                            .SingleOrDefault(iType => iType.IsGenericType && iType.GetGenericTypeDefinition() == typeof(IEnumerable<>));
+                        if (iEnumerableDefinition != null)
```
Comment (jnyrup) [inline Src/FluentAssertions/Types/TypeSelector.cs:248]: nit: please add an empty line after the multiline statement.

File: Src/FluentAssertions/Types/TypeSelector.cs [hunk 11]
```diff
@@ -165,6 +167,96 @@ public TypeSelector ThatAreNotUnderNamespace(string @namespace)
...
+                {
+                    if (type.IsGenericType)
+                    {
+                        if (type.GetGenericTypeDefinition() == typeof(IEnumerable<>))
+                        {
+                            return type.GetGenericArguments().Single();
+                        }
+
+                        var iEnumerableDefinition = type.GetInterfaces()
+                            .SingleOrDefault(iType => iType.IsGenericType && iType.GetGenericTypeDefinition() == typeof(IEnumerable<>));
```
Comment (jnyrup) [inline Src/FluentAssertions/Types/TypeSelector.cs:247]: This might make too strong assumptions about the user's code.

E.g. if a type has multiple implementations of `IEnumerable<>` this throws
```c#
public class C : IEnumerable<int>, IEnumerable<string>
{
    private readonly IEnumerable<int> integers;
    private readonly IEnumerable<string> strings;

    public IEnumerator<int> GetEnumerator() => integers.GetEnumerator();
    
    IEnumerator IEnumerable.GetEnumerator() => ((IEnumerable)integers).GetEnumerator();
    
    IEnumerator<string> IEnumerable<string>.GetEnumerator() => strings.GetEnumerator();
}
```

```
System.InvalidOperationException: 'Sequence contains more than one matching element'
```

File: Tests/FluentAssertions.Specs/Types/MethodInfoSelectorAssertionSpecs.cs [hunk 12]
```diff
@@ -231,5 +232,109 @@ public void When_asserting_methods_are_not_decorated_with_attribute_but_they_are
                              "*ClassWithAllMethodsDecoratedWithDummyAttribute.ProtectedDoNothing*" +
                              "*ClassWithAllMethodsDecoratedWithDummyAttribute.PrivateDoNothing");
         }
+
+
```
Comment (jnyrup) [inline Tests/FluentAssertions.Specs/Types/MethodInfoSelectorAssertionSpecs.cs:236]: nit: multiple empty lines

File: Tests/FluentAssertions.Specs/Types/MethodInfoSelectorSpecs.cs [hunk 13]
```diff
@@ -308,6 +324,14 @@ internal class TestClassForMethodSelectorWithNonInheritableAttributeDerived : Te
...
     }
 
+    internal class TestClassForMethodReturnTypesSelector
+    {
+        public void SomeMethod() { }
+        public int AnotherMethod() { return default; }
+        public string OneMoreMethod() { return default; }
+    }
+
+
```
Comment (jnyrup) [inline Tests/FluentAssertions.Specs/Types/MethodInfoSelectorSpecs.cs:334]: nit: double empty line

File: Tests/FluentAssertions.Specs/Types/TypeSelectorSpecs.cs [hunk 14]
```diff
@@ -510,6 +650,43 @@ internal class SomeCommonClass
     }
 }
 
+namespace Internal.NotOnlyClasses.Test
+{
+    internal class NotOnlyClassesClass
+    {
+    }
+    internal enum NotOnlyClassesEnumeration
```
Comment (jnyrup) [inline Tests/FluentAssertions.Specs/Types/TypeSelectorSpecs.cs:658]: Please add an empty line between type declarations

File: Tests/FluentAssertions.Specs/Types/TypeSelectorSpecs.cs [hunk 15]
```diff
@@ -510,6 +650,43 @@ internal class SomeCommonClass
...
+{
+    internal class ClassToExploreItsMethodsReturnValues
+    {
+        internal void Do() { }
+        internal int DoWithInt() { return default; }
+        internal Task DoWithTask() { return Task.CompletedTask; }
+        internal Task<bool> DoWithTaskBoolean() { return Task.FromResult(true); }
+        internal List<string> DoWithListOfString() { return default; }
+        internal IEnumerable<long> DoWithEnumerableOfBooleans() { return default; }
+
```
Comment (jnyrup) [inline Tests/FluentAssertions.Specs/Types/TypeSelectorSpecs.cs:686]: extra newline

File: docs/_pages/releases.md [hunk 16]
```diff
@@ -21,6 +21,18 @@ sidebar:
 * Make `DefaultValueFormatter` and `EnumerableValueFormatter` suitable for inheritance - [#1295](https://github.com/fluentassertions/fluentassertions/pull/1295).
 * Added support for dictionary assertions on `IReadOnlyDictionary<TKey, TValue>` - [#1298](https://github.com/fluentassertions/fluentassertions/pull/1298).
 * `GenericAsyncFunctionAssertions` now has `AndWhichConstraint` overloads for `NotThrow[Async]` and `NotThrowAfter[Async]` - [#1289](https://github.com/fluentassertions/fluentassertions/pull/1289).
+* Added `ReturnTypes` to `MethodInfoSelector` to get all return types from all the methods selected
```
Comment (jnyrup) [inline docs/_pages/releases.md:24]: This is a nice contribution, but I don't think it needs 12 lines of mentioning ;)
We could condense this some, without loosing information.
E.g. 
"Added `ThatAre[Not]Classes` and `ThatAre[Not]StaticClasses` to `TypeSelector`"

File: CONTRIBUTING.md [hunk 17]
```diff
@@ -5,4 +5,4 @@ No open-source project is going to be successful without contributions. After we
 * The [Pull Request](https://help.github.com/articles/using-pull-requests) is targeted at the `develop` branch.
 * The code complies with the [Coding Guidelines for C# 3.0, 4.0 and 5.0](https://csharpcodingguidelines.com/)/.
 * The changes are covered by a new or existing set of unit tests which follow the Arrange-Act-Assert syntax such as is used [in this example](https://github.com/fluentassertions/fluentassertions/blob/daaf35b9b59b622c96d0c034e8972a020b2bee55/Tests/FluentAssertions.Shared.Specs/BasicEquivalencySpecs.cs#L33).
-* If the contribution affects the documentation, please update [**the documentation**](https://github.com/fluentassertions/fluentassertions/tree/master/docs), which is published on the [website](https://fluentassertions.com/documentation/).
+* If the contribution affects the documentation, please update [**the documentation**](https://github.com/fluentassertions/fluentassertions/tree/master/docs), which is published on the [website](https://fluentassertions.com/introduction/).
```
Comment (jnyrup) [inline CONTRIBUTING.md:8]:
```suggestion
* If the contribution affects the documentation, please update [**the documentation**](https://github.com/fluentassertions/fluentassertions/tree/master/docs), which is published on the [website](https://fluentassertions.com/introduction).
```

File: Src/FluentAssertions/TypeEnumerableExtensions.cs (hunk 7 above)
Comment (fshchudlo) [reply Src/FluentAssertions/TypeEnumerableExtensions.cs:101]: This is the place where I was also taken aback. 
What do you think, the name `That [Not] AreStatic` is not confusing, as if it were about static methods and properties?

Comment (fshchudlo) [reply Src/FluentAssertions/TypeEnumerableExtensions.cs:101]: BTW, is a Type Extensions. I'll name it as you suppose.

Comment (fshchudlo) [reply Src/FluentAssertions/TypeEnumerableExtensions.cs]: Done

File: Src/FluentAssertions/Types/TypeSelector.cs (hunk 9 above)
Comment (fshchudlo) [reply Src/FluentAssertions/Types/TypeSelector.cs:220]: Yep, done

File: Src/FluentAssertions/Types/TypeSelector.cs (hunk 10 above)
Comment (fshchudlo) [reply Src/FluentAssertions/Types/TypeSelector.cs]: Fixed

File: Src/FluentAssertions/Types/TypeSelector.cs (hunk 11 above)
Comment (fshchudlo) [reply Src/FluentAssertions/Types/TypeSelector.cs]: Done

File: Tests/FluentAssertions.Specs/Types/MethodInfoSelectorSpecs.cs (hunk 13 above)
Comment (fshchudlo) [reply Tests/FluentAssertions.Specs/Types/MethodInfoSelectorSpecs.cs]: Fixed

File: Tests/FluentAssertions.Specs/Types/TypeSelectorSpecs.cs (hunk 14 above)
Comment (fshchudlo) [reply Tests/FluentAssertions.Specs/Types/TypeSelectorSpecs.cs:644]: Fixed

File: Tests/FluentAssertions.Specs/Types/TypeSelectorSpecs.cs (hunk 15 above)
Comment (fshchudlo) [reply Tests/FluentAssertions.Specs/Types/TypeSelectorSpecs.cs]: Fixed

File: docs/_pages/releases.md (hunk 16 above)
Comment (fshchudlo) [reply docs/_pages/releases.md:24]: Reduced :)

File: Src/FluentAssertions/TypeEnumerableExtensions.cs [hunk 18]
```diff
@@ -77,5 +77,60 @@ public static IEnumerable<Type> ThatImplement<T>(this IEnumerable<Type> types)
...
+        /// <summary>
+        /// Filters to only include types that satisfies the <paramref name="predicate"/> passed.
+        /// </summary>
+        public static IEnumerable<Type> ThatAre(this IEnumerable<Type> types, Func<Type, bool> predicate)
+        {
+            return new TypeSelector(types).ThatAre(predicate);
+        }
+
+        /// <summary>
+        /// Returns T for the types which are Task&lt;T&gt; or ValueTask&lt;T&gt;; the type itself otherwise
```
Comment (jnyrup) [inline Src/FluentAssertions/TypeEnumerableExtensions.cs:121]:
```suggestion
        /// Returns T for the types which are <see cref="Task{T}" /> or <see cref="ValueTask{T}" />; the type itself otherwise
```
The same for the other uses of `&lt;` in this PR.

File: Src/FluentAssertions/Types/TypeSelector.cs [hunk 19]
```diff
@@ -165,6 +166,106 @@ public TypeSelector ThatAreNotUnderNamespace(string @namespace)
...
+            {
+                if (type.IsGenericType && type.GetGenericTypeDefinition() == typeof(Task<>))
+                {
+                    return type.GetGenericArguments().Single();
+                }
+                if (type.IsGenericType && type.GetGenericTypeDefinition() == typeof(ValueTask<>))
+                {
+                    return type.GetGenericArguments().Single();
+                }
+                return type == typeof(Task) ? typeof(void) : type;
```
Comment (jnyrup) [inline Src/FluentAssertions/Types/TypeSelector.cs:228]: The non-generic `ValueTask` should also return `void`

File: Src/FluentAssertions/Types/TypeSelectorAssertions.cs [hunk 20]
```diff
@@ -292,6 +292,47 @@ public AndConstraint<TypeSelectorAssertions> NotBeDecoratedWithOrInherit<TAttrib
             return new AndConstraint<TypeSelectorAssertions>(this);
         }
 
+        /// <summary>
+        /// Asserts that the all <see cref="Type"/> are sealed classes
```
Comment (jnyrup) [inline Src/FluentAssertions/Types/TypeSelectorAssertions.cs:296]: `TypeSelectorAssertions` has singular/plural mixed up.
`PropertyInfoSelectorAssertions.cs` and `MethodInfoSelectorAssertions.cs` use plural form.

```suggestion
        /// Asserts that the selected types are sealed
```

Feel free to fix the summaries for the other methods in this class, otherwise I'll fix them later :)

File: Src/FluentAssertions/TypeEnumerableExtensions.cs [hunk 21]
```diff
@@ -77,5 +77,60 @@ public static IEnumerable<Type> ThatImplement<T>(this IEnumerable<Type> types)
...
+        }
+
+        /// <summary>
+        /// Filters to only include types that are not classes.
+        /// </summary>
+        public static IEnumerable<Type> ThatAreNotClasses(this IEnumerable<Type> types)
+        {
+            return new TypeSelector(types).ThatAreNotClasses();
+        }
+        /// <summary>
```
Comment (jnyrup) [inline Src/FluentAssertions/TypeEnumerableExtensions.cs:96]: Missing empty line.
There are also other cases, but I'll leave it as an exercise for the reader to find them.

File: Src/FluentAssertions/TypeEnumerableExtensions.cs [hunk 22]
```diff
@@ -77,5 +77,60 @@ public static IEnumerable<Type> ThatImplement<T>(this IEnumerable<Type> types)
...
+        /// </summary>
+        public static IEnumerable<Type> ThatAreNotStatic(this IEnumerable<Type> types)
+        {
+            return new TypeSelector(types).ThatAreNotStatic();
+        }
+
+        /// <summary>
+        /// Filters to only include types that satisfies the <paramref name="predicate"/> passed.
+        /// </summary>
+        public static IEnumerable<Type> ThatAre(this IEnumerable<Type> types, Func<Type, bool> predicate)
```
Comment (jnyrup) [inline Src/FluentAssertions/TypeEnumerableExtensions.cs:115]: Thinking aloud.
Would it be better to name this `ThatSatisfies`?
@dennisdoomen 

```c#
types.ThatAre(t => t.IsCSharpStatic())
types.ThatSatisfies(t => t.IsCSharpStatic())
```

File: Src/FluentAssertions/Types/MethodInfoSelector.cs (hunk 3 above)
Comment (jnyrup) [inline Src/FluentAssertions/Types/MethodInfoSelector.cs:135]: Could this be added to `PropertyInfoSelector` as well for consistency?

File: Src/FluentAssertions/Types/TypeSelector.cs [hunk 23]
```diff
@@ -165,6 +166,106 @@ public TypeSelector ThatAreNotUnderNamespace(string @namespace)
             return this;
         }
 
+        /// <summary>
+        /// Determines whether the type is class
```
Comment (jnyrup) [inline Src/FluentAssertions/Types/TypeSelector.cs:170]:
```suggestion
        /// Determines whether the type is a class
```

File: Src/FluentAssertions/TypeEnumerableExtensions.cs (hunk 18 above)
Comment (fshchudlo) [reply Src/FluentAssertions/TypeEnumerableExtensions.cs]: Fixed

File: Src/FluentAssertions/Types/TypeSelector.cs (hunk 19 above)
Comment (fshchudlo) [reply Src/FluentAssertions/Types/TypeSelector.cs:228]: Fixed

File: Src/FluentAssertions/TypeEnumerableExtensions.cs (hunk 21 above)
Comment (fshchudlo) [reply Src/FluentAssertions/TypeEnumerableExtensions.cs:96]: Fixed

File: Src/FluentAssertions/Types/MethodInfoSelector.cs (hunk 3 above)
Comment (fshchudlo) [reply Src/FluentAssertions/Types/MethodInfoSelector.cs:135]: Done

File: Src/FluentAssertions/TypeEnumerableExtensions.cs (hunk 22 above)
Comment (fshchudlo) [reply Src/FluentAssertions/TypeEnumerableExtensions.cs:116]: In my opinion `ThatAre((t => t.IsCSharpStatic())` looks more fluid. But this is up to you.

Comment (dennisdoomen) [reply Src/FluentAssertions/TypeEnumerableExtensions.cs:116]: IMHO, I think `ThatSatisfy` is even better.

Comment (fshchudlo) [reply Src/FluentAssertions/TypeEnumerableExtensions.cs]: You're right. This is the best option. Done.

Comment (jnyrup) [reply Src/FluentAssertions/TypeEnumerableExtensions.cs]: phew, thanks for catching my plural typo.
This also aligns better with the `SatisfyRespectively` name.

**Comment by jnyrup** [comment]
@fshchudlo Thanks for the contribution!

**Comment by fshchudlo** [comment]
@jnyrup I'm sure this is my first contribution, but not the last :) Thank you for such a thorough study of PR and such a cool library :)
